### PR TITLE
Swap predicates for updateable optional values in 'ensure' and 'validation'

### PR DESCRIPTION
**Pull request by emilypi**
Discussion here: #32 

Also, found some unused imports

### CODE REVIEW

**Comment by gwils** [comment]
Please update the tests as well.

**Comment by emilypi** [comment]
ah these tests weren't caught by building `all`. My bad.

File: src/Data/Validation.hs [hunk 1]
```diff
@@ -203,16 +202,18 @@ instance (NFData e, NFData a) => NFData (Validation e a) where
...
 --
 -- @
--- validate :: e -> (a -> Bool) -> a -> Validation e a
+-- validate :: e -> (a -> Maybe b) -> a -> Validation e b
 -- @
-validate :: Validate v => e -> (a -> Bool) -> a -> v e a
-validate e p a =
-  if p a then _Success # a else _Failure # e
+validate :: Validate v => e -> (a -> Maybe b) -> a -> v e b
+validate e p a = case p a of
```
Comment (FintanH) [inline src/Data/Validation.hs]: Why not `maybe (_Failure # b) ((#) _Success) . p`? Just out of interest, I'm only a tourist in this repo :blush:

Comment (emilypi) [reply src/Data/Validation.hs]: I did think about this variation first, but I decided against it doesn't bring much to the code. It doesn't pass my personal golfing threshold of "is this more elegant than a pattern match?", and it also wasn't idiomatic when compared to the rest of the code.

Comment (FintanH) [reply src/Data/Validation.hs]: That's fair! I'd go for a birdie and write `maybe (_Failure # e) (\b -> _Success # b) . p` but this version is also readable and elegant :100:

Comment (emilypi) [reply src/Data/Validation.hs]: a birdie 😆

**Pull request merged**